From 0000000000000000000000000000000000000000 Mon Sep 17 00:00:00 2001
From: rmp22 <195054967+rmp22@users.noreply.github.com>
Date: Fri, 1 May 2026 06:12:49 +0800
Subject: [PATCH 226/226] services: Optimizing wms prop | settings reads
MIME-Version: 1.0
Content-Type: text/plain; charset=UTF-8
Content-Transfer-Encoding: 8bit

the props and settings reads were being done at every window iteration - expensive, add an observer and add caching mechanism

Change-Id: I5886612fe1e732d4c23043933b164097110c50c9
Signed-off-by: Ghosuto <clash.raja10@gmail.com>


[MikolajQ: konflikt rozwiązany przy przenoszeniu na LineageOS lineage-23.2]
---
 .../server/wm/WindowManagerService.java       | 44 ++++++++++++-------
 1 file changed, 27 insertions(+), 17 deletions(-)

diff --git a/services/core/java/com/android/server/wm/WindowManagerService.java b/services/core/java/com/android/server/wm/WindowManagerService.java
index c4fd7cad4f..af7cc81ff6 100644
--- a/services/core/java/com/android/server/wm/WindowManagerService.java
+++ b/services/core/java/com/android/server/wm/WindowManagerService.java
@@ -446,8 +446,10 @@ public class WindowManagerService extends IWindowManager.Stub
     /** Indicates we are removing the focused window when updating the focus. */
     static final int UPDATE_FOCUS_REMOVING_FOCUS = 4;
 
-    private static final String SYSTEM_SECURE = "ro.secure";
-    private static final String SYSTEM_DEBUGGABLE = "ro.debuggable";
+    private static final boolean SYSTEM_SECURE = SystemProperties.getBoolean("ro.secure", true);
+    private static final boolean SYSTEM_DEBUGGABLE =
+            SystemProperties.getBoolean("ro.debuggable", false);
+    private static final String WINDOW_IGNORE_SECURE = "window_ignore_secure";
 
     private static final String DENSITY_OVERRIDE = "ro.config.density_override";
     private static final String SIZE_OVERRIDE = "ro.config.size_override";
@@ -835,6 +837,8 @@ public class WindowManagerService extends IWindowManager.Stub
                 Settings.Secure.getUriFor(Settings.Secure.IMMERSIVE_MODE_CONFIRMATIONS);
         private final Uri mDisableSecureWindowsUri =
                 Settings.Secure.getUriFor(Settings.Secure.DISABLE_SECURE_WINDOWS);
+        private final Uri mWindowIgnoreSecureUri =
+                Settings.Secure.getUriFor(WINDOW_IGNORE_SECURE);
         private final Uri mMagnifyImeEnabledUri = Settings.Secure.getUriFor(
                 Settings.Secure.ACCESSIBILITY_MAGNIFICATION_MAGNIFY_NAV_AND_IME);
         private final Uri mPolicyControlUri =
@@ -869,6 +873,8 @@ public class WindowManagerService extends IWindowManager.Stub
                     UserHandle.USER_ALL);
             resolver.registerContentObserver(mDisableSecureWindowsUri, false, this,
                     UserHandle.USER_ALL);
+            resolver.registerContentObserver(mWindowIgnoreSecureUri, false, this,
+                    UserHandle.USER_ALL);
             if (com.android.server.accessibility.Flags.enableMagnificationMagnifyNavBarAndIme()) {
                 resolver.registerContentObserver(mMagnifyImeEnabledUri, false, this,
                         UserHandle.USER_ALL);
@@ -929,7 +935,7 @@ public class WindowManagerService extends IWindowManager.Stub
                 return;
             }
 
-            if (mDisableSecureWindowsUri.equals(uri)) {
+            if (mDisableSecureWindowsUri.equals(uri) || mWindowIgnoreSecureUri.equals(uri)) {
                 updateDisableSecureWindows();
                 return;
             }
@@ -1051,23 +1057,26 @@ public class WindowManagerService extends IWindowManager.Stub
         }
 
         void updateDisableSecureWindows() {
-            if (!SystemProperties.getBoolean(SYSTEM_DEBUGGABLE, false)) {
-                return;
-            }
-
-            boolean disableSecureWindows;
-            try {
-                disableSecureWindows = Settings.Secure.getIntForUser(mContext.getContentResolver(),
-                        Settings.Secure.DISABLE_SECURE_WINDOWS, 0) != 0;
-            } catch (Settings.SettingNotFoundException e) {
-                disableSecureWindows = false;
+            final ContentResolver resolver = mContext.getContentResolver();
+            final boolean ignoreSecureWindows =
+                    Settings.Secure.getInt(resolver, WINDOW_IGNORE_SECURE, 0) == 1;
+            boolean disableSecureWindows = false;
+            if (SYSTEM_DEBUGGABLE) {
+                try {
+                    disableSecureWindows = Settings.Secure.getIntForUser(resolver,
+                            Settings.Secure.DISABLE_SECURE_WINDOWS, 0) != 0;
+                } catch (Settings.SettingNotFoundException e) {
+                    disableSecureWindows = false;
+                }
             }
-            if (mDisableSecureWindows == disableSecureWindows) {
+            if (mDisableSecureWindows == disableSecureWindows
+                    && mIgnoreSecureWindows == ignoreSecureWindows) {
                 return;
             }
 
             synchronized (mGlobalLock) {
                 mDisableSecureWindows = disableSecureWindows;
+                mIgnoreSecureWindows = ignoreSecureWindows;
                 mRoot.refreshSecureSurfaceState();
             }
         }
@@ -1247,6 +1256,7 @@ public class WindowManagerService extends IWindowManager.Stub
     private final ScreenRecordingCallbackController mScreenRecordingCallbackController;
 
     private volatile boolean mDisableSecureWindows = false;
+    private volatile boolean mIgnoreSecureWindows = false;
 
     /** Creates an instance of the WindowManagerService for the system server. */
     public static WindowManagerService main(@NonNull final Context context,
@@ -5466,8 +5476,7 @@ public class WindowManagerService extends IWindowManager.Stub
     }
 
     private boolean isSystemSecure() {
-        return "1".equals(SystemProperties.get(SYSTEM_SECURE, "1")) &&
-                "0".equals(SystemProperties.get(SYSTEM_DEBUGGABLE, "0"));
+        return SYSTEM_SECURE && !SYSTEM_DEBUGGABLE;
     }
 
     /**
@@ -7275,6 +7284,7 @@ public class WindowManagerService extends IWindowManager.Stub
         });
         pw.print("  mBlurEnabled="); pw.println(mBlurController.getBlurEnabled());
         pw.print("  mDisableSecureWindows="); pw.println(mDisableSecureWindows);
+        pw.print("  mIgnoreSecureWindows="); pw.println(mIgnoreSecureWindows);
 
         mInputManagerCallback.dump(pw, "  ");
         mSnapshotController.dump(pw, " ");
@@ -10881,7 +10891,7 @@ public class WindowManagerService extends IWindowManager.Stub
     }
 
     boolean getDisableSecureWindows() {
-        return mDisableSecureWindows;
+        return mIgnoreSecureWindows || mDisableSecureWindows;
     }
 
     /**
